Breakup main tasks into mockable subroutines, support "v1" or "v2" schema per https://github.com/cisagov/pen-test-portal/pull/3

S4lt5 opened the pull request:
Continued from #51 

## 🗣 Description ##

To address #15, some of the major actions of the task are broken up into subtasks, and try to give a descriptive outer exception on what went wrong, while passing a trace and inner exception to preserve the "actual" error.

Additionally, this now supports "v1" and "v2" schemas as described @ https://github.com/cisagov/pen-test-portal/pull/3

## 💭 Motivation and context ##

I think this helps with readability and testability, since unit tests can now mock up e.g. the s3 test or test them against ephemeral services to make sure things are working correctly.

This is going to be necessary to adapt the changes from the pen-test-portal project as currently testing this task outside of it's native environment is difficult.


## 🧪 Testing ##
The new validation and update functions have test coverage now where none existed before.

Mocked database calls are checked to see if the main logic is correctly updating the back end database.

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass. _(pre commit hook does not pass, but it succeeds more than it did before!)_
- [ ] @mcdonnnj will deploy this to his testing environment and use the sample JSONs from https://github.com/cisagov/pen-test-portal/pull/3 and verify expected functionality.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been approved. -->

- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Add a tag or create a release.

S4lt5 committed:
Cut out bits no longer relevant to current lambda format. Add expected test artifacts

S4lt5 committed:
Rework changes from old PR, and check tests still pass

S4lt5 committed:
formatting changes for linter.

S4lt5 committed:
more linting..

S4lt5 committed:
remove inadvertent duplicate download call, fix flake8 lint issues

S4lt5 committed:
flake8 fixes

S4lt5 committed:
Add empty package def to satisfy pre commit hooks and allow imports for testing.

Fix various linting issues.

S4lt5 committed:
Fix markdown lint issues

S4lt5 committed:
Revert README to develop branch

S4lt5 committed:
Add bit about default test field mapping file to docs

S4lt5 committed:
Fix Json with pretty-format-json

S4lt5 committed:
try tests within src dir for relative import

S4lt5 committed:
Add missing requirements to project

jsf9k commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
+boto3==1.26.41
+pymongo==4.3.3
```
We don't usually pin this hard in the `requirements*.txt` files unless there is a real need to do so, in which case there would also need to be a comment explaining the reason.

S4lt5 replied:
Got it. I was trying to avoid https://github.com/cisagov/findcdn/issues/54 which bit me in the ass recently!
I'll change it to >= current minor revision.

jsf9k replied:
Addressed in commit f006e82.

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,288 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    Parameters:
+    -----------
+    s3_client : S3.Client
+        The AWS S3 Client to retrieve the data file with .
```
```suggestion
        The AWS S3 Client to retrieve the data file with.
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,288 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    Parameters:
+    -----------
+    s3_client : S3.Client
+        The AWS S3 Client to retrieve the data file with .
```
```suggestion
        The AWS S3 Client to retrieve the data file with.
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,288 @@ def move_processed_file(s3_client, bucket, folder, filename):
+
+    Returns
+    -------
+        string, : The file path of the newly created temp file
```
I think there is an extra comma here:
```suggestion
        string : The file path of the newly created temp file
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,288 @@ def move_processed_file(s3_client, bucket, folder, filename):
+                    finding[field_map_dict[field]] = finding[field]
+                finding.pop(field, None)
+
+        # work with v1 and v2. If has NCATS ID  OR findings the document is probably OK
```
```suggestion
        # work with v1 and v2. If has NCATS ID OR findings the document is probably OK
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,288 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        The database to update
+
+    finding: dict
+        The finding data to insert.1
```
```suggestion
        The finding data to insert.
```

jsf9k reviewed (changes requested):
Is your test code meant to be in `src/tests` or `tests`?  You seem to have the same files in both locations.

Also, we strongly encourage the use of the old-style format strings over f-strings when calling `log.debug()`, etc.  The reason is that if you use the old-style format strings then the formatting of the string will be deferred by the logging library until absolutely necessary and will not be done at all if the log message will not be output by any handlers.  The efficiency savings are worth using the more cumbersome format and making this a best practice for log messages.  This topic is discussed [here](https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs) and countless other places.

S4lt5 committed:
Update src/findings_data_import.py

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

S4lt5 committed:
Update src/findings_data_import.py

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

S4lt5 committed:
Update src/findings_data_import.py

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

S4lt5 committed:
Update src/findings_data_import.py

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

S4lt5 committed:
Update src/findings_data_import.py

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

S4lt5 committed:
Update requirements.txt

Try more permissive version pinning

S4lt5 commented:
> Is your test code meant to be in `src/tests` or `tests`? You seem to have the same files in both locations.
> 
> Also, we strongly encourage the use of the old-style format strings over f-strings when calling `log.debug()`, etc. The reason is that if you use the old-style format strings then the formatting of the string will be deferred by the logging library until absolutely necessary and will not be done at all if the log message will not be output by any handlers. The efficiency savings are worth using the more cumbersome format and making this a best practice for log messages. This topic is discussed [here](https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs) and countless other places.



> Also, we strongly encourage the use of the old-style format strings over f-strings when calling `log.debug()`, etc. The reason is that if you use the old-style format strings then the formatting of the string will be deferred by the logging library until absolutely necessary and will not be done at all if the log message will not be output by any handlers. The efficiency savings are worth using the more cumbersome format and making this a best practice for log messages. This topic is discussed [here](https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs) and countless other places.

The duplication of tests is indeed unintentional, originally had them in the tests/ path but had issues with mypy and referencing fdi from two different locations. 

As far as the format-strings, understood and I'll make the changes today.

jsf9k commented:
In the future, when you accept someone's suggester changes, please make the comment associated with the commit something more specific and more informative than "Update src/findings_data_import.py".  For example, you could have grouped several of my suggestions together in a single batch (see below) with a comment like, "Remove extra spaces" or something similar.  Even "Fix a typo" is better than "Update src/findings_data_import.py" because then you know at a glance that the commit is probably not a substantive one.

(You can batch suggestions together into a single commit if you view them on the "Files changed" tab of the pull request.)

Commit:
Remove redunant tests now in src/tests

Commit:
Use lazy % string formatting on logger calls.

Initialize temp_data_filepath in import_data in case an exception is thrown before it is initalized, only delete a file if it is created first..

Commit:
Un-clobber original log format: "{valid}/{total}" vs "{valid/total}"

S4lt5 commented:
> In the future, when you accept someone's suggester changes, please make the comment associated with the commit something more specific and more informative than "Update src/findings_data_import.py". For example, you could have grouped several of my suggestions together in a single batch (see below) with a comment like, "Remove extra spaces" or something similar. Even "Fix a typo" is better than "Update src/findings_data_import.py" because then you know at a glance that the commit is probably not a substantive one.
> 
> (You can batch suggestions together into a single commit if you view them on the "Files changed" tab of the pull request.)

Gotcha.. not really a huge user of the web UI but figured I would give it a whirl here. We didn't have such features in our on-site repo in the last gig!

S4lt5 committed:
Include reformatting by https://github.com/psf/black

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    db_hostname : str
+        The hostname that has the database to store the data in.
+
+    db_port : str
+        The port that the database server is listening on. [default: 27017]
```
Why wouldn't this be an `int`?

S4lt5 replied:
Because, inexplicably, it already is in the current production branch and I didn't want to make any more waves than already accomplished

The current code uses a hard-coded port string explicitly, and documents it as a string. And really, since we're working with python the typing is more of a suggestion than anything else.

https://github.com/cisagov/findings-data-import-lambda/blob/2c3ed77281388897a149e60c278a42652a40c367/src/findings_data_import.py#L60-L87

To be clear, this is not something I changed, and it was not relevant to the changes I was making.

jsf9k replied:
Fair enough.

jsf9k replied:
> And really, since we're working with python the typing is more of a suggestion than anything else.

That is kind of true, but it's always good to document any constraints you are expecting on input variables.  In this case `db_port` is pretty clearly expected to be an `int` (or at least a `str` that can be parsed into an `int`).

S4lt5 replied:
I don't disagree at all, it should be an int unless there is an explicit reason. I mostly just want to limit the scope of this change where possible, but can edit it here in this PR as well if need be.

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            )
+            db_info[key] = response["Parameter"]["Value"]
+
+        logging.info("Connecting to the mongo db at %s %s", db_hostname, db_port)
```
If you change `db_port` to an `int`, then this should change too:
```suggestion
        logging.info("Connecting to the mongo db at %s %d", db_hostname, db_port)
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        )
+        db = db_connection[db_info["db_name"]]
+        logging.info(
+            "DB connection set up to %s:%s/%s", db_hostname, db_port, db_info["db_name"]
```
If you change `db_port` to an `int` then this should change too:
```suggestion
            "DB connection set up to %s:%d/%s", db_hostname, db_port, db_info["db_name"]
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            and "findings" not in finding.keys()
+        ):
+            logging.warning(
+                "Skipping record %s. Missing 'RVA ID' or 'NCATS ID' field.", index
```
`index` is an `int`, right?
```suggestion
                "Skipping record %d. Missing 'RVA ID' or 'NCATS ID' field.", index
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            finding["RVA ID"] = rID
+        else:
+            logging.warning(
+                "Skipping record %s: Unable to extract valid RVA ID from '%'",
```
`index` is an `int`, right?  Also, is there an `s` missing here?
```suggestion
                "Skipping record %d: Unable to extract valid RVA ID from '%s'",
```

S4lt5 replied:
It is an int, in general I don't try to get too picky with python format strings mostly because showing me the str() representation of any particular value will give me what i want when I'm showing a number, and outright fail less times when I'm getting something unexpected. But I concede there's really no reason to keep it a string so I'll make the change.

jsf9k commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        valid_findings.append(finding)
+
+    logging.info(
+        "{%s/%s} documents successfully processed.",
```
Both variables are integers:
```suggestion
        "{%d/%d} documents successfully processed.",
```

jsf9k commented on src/findings_data_import.py:
```diff
@@ -132,103 +408,35 @@ def import_data(
         logging.info(
-            '%d/%d documents successfully processed from "%s"',
-            processed_findings,
+            "%s/%s documents successfully processed from '%s'.",
```
I think the first two variables are always `int`s:
```suggestion
            "%d/%d documents successfully processed from '%s'.",
```

jsf9k reviewed (changes requested):
Some fine-tuning of the logging calls is necessary.

S4lt5 committed:
Apply suggestions from code review

Change format string specifiers where it makes sense, left port because it's explicitly a string and documented as such, and changing that is sorta out of scope for what I need to do here.

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

mcdonnnj commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
+boto3>=1.26
+pymongo>=4.3
```
```suggestion
```

S4lt5 replied:
I don't understand the above, do these get installed in some other manner during build?

mcdonnnj replied:
They only need to be installed because of your testing code. These requirements files are only used for local environment setup. If you look in a [Pipfile](https://github.com/cisagov/findings-data-import-lambda/blob/2c3ed77281388897a149e60c278a42652a40c367/src/py3.7/Pipfile#L9-L12) you'll see they are listed as dependencies. A given Pipfile (and its corresponding lock file) are what is used to generate the Lambda Docker image and deployment artifact.

mcdonnnj commented on src/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Findings data import package that covers file retrieval, parsing, validation, and database update."""
```
Delete this file please.

mcdonnnj commented on src/tests/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Basic tests of findings-data-import to ensure filtering and updating is occuring as expected."""
```
Delete this file please.

mcdonnnj commented on src/tests/artifacts/expected_dhs_v1.json:
```diff
@@ -0,0 +1,250 @@
+[
```
Delete this file please.

mcdonnnj commented on src/tests/artifacts/expected_dhs_v2.json:
```diff
@@ -0,0 +1,378 @@
+{
```
Delete this file please.

mcdonnnj commented on src/tests/artifacts/field_map.json:
```diff
@@ -0,0 +1,9 @@
+{
```
Delete this file please.

mcdonnnj commented on src/tests/test_findings_data_import.py:
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env pytest -vs
```
Delete this file please.

mcdonnnj commented on README.md:
```diff
@@ -27,6 +27,12 @@ script will remove that element from the record.
 
+## Default Field Mapping File ##
+
+A default mapping file exists at `tests/artifacts/field_map.json` and should be
+used as a starting point to be compatible with most versions of the Pen Test Portal's
+findings JSON output.
```
Please provide an example here in the README instead. That is our common practice as seen in [cisagov/cool-assessment-terraform](https://github.com/cisagov/cool-assessment-terraform#pre-requisites) for an appropriate Terraform variables file.

S4lt5 replied:
> > > I would like the testing code added here removed. The newer lambda structure is based on creating a deployment and not a Python package so turning things into Python modules just for testing is not a workaround I can support.
> > 
> > 
> > Would you suggest something like the following, if @Yablargo wants to preserve the testing code?
> > ```
> > 1. Create a separate Python library based on the [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library) skeleton that handles the heavy lifting.
> > 
> > 2. Add the mock and test code there so that the heavy lifting is tested.
> > 
> > 3. Leverage that Python library in this repo.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This does align with how I have done Lambda testing in the past, but I wanted to make sure you didn't have something else in mind.
> 
> Yes if the Python code is complicated enough it should be broken into its own Python project that is added as a dependency in the `Pipfile`s for the Lambda configuration. These Lambda projects should be designed around creating deployment artifacts and not storing a lot of logic. The only reason the logic remains in this project is because of timelines around getting this change out. Starting from scratch it is not how I would implement this functionality.

I guess this can serve my formal complaint on the matter, I think it is quite absurd to have the logic in this sort of project, and then claim that no testing is preferable to making a good faith attempt to improve testing because whoever wrote this should not have done so, years ago.

That said, I'll remove it if everyone agrees that we're better off with zero tests on things, although I'd suggest that we probably need a lot more across all of these projects since something as simple as package updates might break these sorts of things in interesting ways.

I guess I'd also say that something like the new lambda structure should have tests as part of the example and it well documented on how an end user is supposed to set up basic functional tests for these things. 

The golden rule **for me** has always been that if you touch functionality you should update tests to cover what you touched. I know that it is still sorely lacking, but the intent was to leave things in a slightly better place than I found them, and I try to use this as a guiding principle when working on any particular codebase.

@jsf9k I guess if you concur with above, I'll just drop all the test code and we can move on with our lives

S4lt5 replied:
I'm gonna open a ticket for future work to add a comprehensive "defaults" to the docs, but I don't really know what that should look like right now and I think it is getting out of scope for this change. I've removed the original "default" text

jsf9k replied:
> @jsf9k I guess if you concur with above, I'll just drop all the test code and we can move on with our lives

I'd honestly be fine with leaving the test code in for now given that #57 exists, but since he is the architect of the Lambda skeleton I'll defer to @mcdonnnj.

I did add an acceptance criterion in #57 to remind us to revive the tests when that work takes place.

S4lt5 replied:
> Please provide an example here in the README instead. That is our common practice as seen in [cisagov/cool-assessment-terraform](https://github.com/cisagov/cool-assessment-terraform#pre-requisites) for an appropriate Terraform variables file.

I just dropped it, this is not really related to the change, and really should be it's own separate PR.

mcdonnnj reviewed (changes requested):
I would like the testing code added here removed. The newer lambda structure is based on creating a deployment and not a Python package so turning things into Python modules just for testing is not a workaround I can support.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
-    """Print a standard message when a record must be skipped."""
-    log_format = 'Skipping record %d of "%s"' + (
-        f": {format_string}" if format_string else ""
+def get_field_map(s3_client=None, s3_bucket=None, field_map=None):
```
Please add type hints to the function signature to reflect the function's docstring.

S4lt5 replied:
I left s3_client as typing.Any since I don't really know how complex type hints work in python, up to some suggestions here.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    ssm_db_name=None,
+    ssm_db_user=None,
+    ssm_db_password=None,
+):
```
Please add type hints to the function signature to reflect the function's docstring.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    s3_client=None,
+    s3_bucket=None,
+    data_filename=None,
+):
```
Please add type hints to the function signature to reflect the function's docstring.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        )
+
+
+def extract_findings(findings_data, field_map_dict):
```
Please add type hints to the function signature to reflect the function's docstring.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    return valid_findings
+
+
+def update_record(db=None, finding=None):
```
Please add type hints to the function signature to reflect the function's docstring.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        raise ValueError("The passed finding was not identifiable as V1 or V2 schema")
 
 
 def import_data(
```
So that things are consistent please add type hints to the function signature to reflect the function's docstring.

S4lt5 replied:
As elsewhere, I'm not really sure how to handle the complex type so used typing.Any for the mongoClient arg

jsf9k commented:
> I would like the testing code added here removed. The newer lambda structure is based on creating a deployment and not a Python package so turning things into Python modules just for testing is not a workaround I can support.

Would you suggest something like the following, if @Yablargo wants to preserve the testing code?
1. Create a separate Python library based on the [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library) skeleton that handles the heavy lifting.
2. Add the mock and test code there so that the heavy lifting is tested.
3. Leverage that Python library in this repo.

This does align with how I have done Lambda testing in the past, but I wanted to make sure you didn't have something else in mind.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        raise Exception(
+            "Unable to decode field map data, does not appear to be valid JSON.",
+            json_err,
+        )
```
I think logging a specific error and re-raising the exception is a better approach.

```suggestion
    except ClientError:
        logging.error(
            "Unable to download the field map data %s from %s", field_map, s3_bucket
        )
        raise
    except json.JSONDecodeError:
        logging.error(
            "Unable to decode field map data, does not appear to be valid JSON."
        )
        raise
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        ).with_traceback()
+    # Handle all mongo exceptions the same way..
+    except Exception as err:
+        raise Exception("Unable to connect to the mongo db", err).with_traceback()
```
Same exception handling rationale as above.

```suggestion
    except ClientError:
        logging.error("Unable to fetch database credentials from the SSM")
        raise
    # Handle all mongo exceptions the same way..
    except Exception:
        logging.error("Unable to connect to the mongo db")
        raise
```

S4lt5 replied:
I added the str(exception) here, figured extra detail is always valuable, but might be overkill.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            Exception(
+                f"Error downloading file {data_filename} from {s3_bucket} ", err
+            ).with_traceback()
+        )
```
Same exception handling rationale as above.

```suggestion
    except json.JSONDecodeError:
        logging.error("Unable to decode JSON data for %s", data_filename)
        raise
    except ClientError:
        logging.error("Error downloading file %s from %s", data_filename, s3_bucket)
        raise
```

mcdonnnj commented:
> > I would like the testing code added here removed. The newer lambda structure is based on creating a deployment and not a Python package so turning things into Python modules just for testing is not a workaround I can support.
> 
> Would you suggest something like the following, if @Yablargo wants to preserve the testing code?
> 
>     1. Create a separate Python library based on the [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library) skeleton that handles the heavy lifting.
> 
>     2. Add the mock and test code there so that the heavy lifting is tested.
> 
>     3. Leverage that Python library in this repo.
> 
> 
> This does align with how I have done Lambda testing in the past, but I wanted to make sure you didn't have something else in mind.

Yes if the Python code is complicated enough it should be broken into its own Python project that is added as a dependency in the `Pipfile`s for the Lambda configuration. These Lambda projects should be designed around creating deployment artifacts and not storing a lot of logic. The only reason the logic remains in this project is because of timelines around getting this change out. Starting from scratch it is not how I would implement this functionality.

Commit:
Remove tests on request

Commit:
Remove unwanted field map sample reference

Commit:
Fix lint requirements.

S4lt5 commented:
In the spirit of moving things along, I've gone ahead and removed the test code, when someone gets around to making this align with your new structure, they can be picked back out if desired.

Commit:
Fix accidental deletion.

mcdonnnj commented:
> In the spirit of moving things along, I've gone ahead and removed the test code, when someone gets around to making this align with your new structure, they can be picked back out if desired.

Please see https://github.com/cisagov/findings-data-import-lambda/issues/57 for this future work.

S4lt5 commented:
> > In the spirit of moving things along, I've gone ahead and removed the test code, when someone gets around to making this align with your new structure, they can be picked back out if desired.
> 
> Please see #57 for this future work.

Looks good to me. I'm a little confused on the requirements.txt however, reading that it appears that since the python script is creating it on the fly it needs the requirements in the script, or no?

mcdonnnj commented:
> > > In the spirit of moving things along, I've gone ahead and removed the test code, when someone gets around to making this align with your new structure, they can be picked back out if desired.
> > 
> > 
> > Please see #57 for this future work.
> 
> Looks good to me. I'm a little confused on the requirements.txt however, reading that it appears that since the python script is creating it on the fly it needs the requirements in the script, or no?

I'm sorry since I'm in the middle of a review it won't let me add a separate comment to reply to your question (it bundles it with my pending review comments). I'm copying it here so you don't have to wait:

> They only need to be installed because of your testing code. These requirements files are only used for local environment setup. If you look in a [Pipfile](https://github.com/cisagov/findings-data-import-lambda/blob/2c3ed77281388897a149e60c278a42652a40c367/src/py3.7/Pipfile#L9-L12) you'll see they are listed as dependencies. A given Pipfile (and its corresponding lock file) are what is used to generate the Lambda Docker image and deployment artifact.

S4lt5 committed:
remove runtime requirements from requirements.txt, as requested

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

S4lt5 commented:
I guess my final comment here is that this was not intended to be a breaking change, and should accept the old data just fine, the only difference being that it will tag it as "v1" but since there was no real control of the various "extra" attributes of a finding object beyond the few mandatory fields, this shouldn't really affect much other than the addition of a new tag which can be safely ignored by consumers.

S4lt5 commented:
Unless I missed something, all outstanding have been resolved, let me know if there's something else to attend to here.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    ssm_db_user=None,
+    ssm_db_password=None,
+):
+    """Set up a mongo db connection based on the supplied host/port and SSM key values.
```
```suggestion
    """Set up a MongoDB connection based on the supplied host/port and SSM key values.
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    s3_bucket=None,
+    data_filename=None,
+):
+    """Download a file from a specified s3 bucket, and place it in a temporary file path.
```
```suggestion
    """Download a file from a specified S3 bucket, and place it in a temporary file path.
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    valid_findings = []
+
+    # if we (v2) get a lone object, wrap it in a list for compatability
+    if type(findings_data) != list:
```
We only care that it is some kind of `list` and not that it is _specifically_ a `list`.

```suggestion
    if not isinstance(findings_data, list):
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    for index, finding in enumerate(findings_data):
+
+        if not finding or not hasattr(finding, "keys"):
+            logging.warning("Received an empty of invalid finding object, skipping.")
```
```suggestion
            logging.warning("Received an empty or invalid finding object, skipping.")
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+    # Iterate through data and save each record to the database
+    for index, finding in enumerate(findings_data):
+
+        if not finding or not hasattr(finding, "keys"):
```
I think this better illustrates that we're looking for dictionaries.

```suggestion
        if not finding or not isinstance(finding, dict):
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            logging.warning("Received an empty of invalid finding object, skipping.")
+            continue
+
+        # Replace or rename fields from replacement JSON
```
```suggestion
        # Replace or rename fields according to the field mapping configuration
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+
+    # if it has "NCATS ID", it is 'v1' record
+    if "NCATS ID" in finding and "Severity" in finding:
+        finding["schema"] = "v1"
```
Since the "v1" keys are all capitalized as far as I know I would prefer to be consistent.

```suggestion
        finding["Schema"] = "v1"
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        The finding data to insert.
+    """
+    if "RVA ID" not in finding:
+        raise ValueError("The passed finding had no RVA ID field.")
```
```suggestion
        raise ValueError("The passed finding has no RVA ID field.")
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -132,103 +408,35 @@ def import_data(
         logging.info(
-            '%d/%d documents successfully processed from "%s"',
-            processed_findings,
+            "%d/%d documents successfully processed from '%s'.",
```
Consistency.

```suggestion
            '%d/%d documents successfully processed from "%s".',
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -114,16 +394,12 @@ def import_data(
     try:
         # This allows us to access keys with spaces in them. When they are passed
-        # in to the lambda the psaces are replaced with plus signs which results
+        # in to the lambda the spaces are replaced with plus signs which results
```
Since we're referring to the AWS Lambda product.

```suggestion
        # in to the Lambda the spaces are replaced with plus signs which results
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+        valid_findings.append(finding)
+
+    logging.info(
+        "{%d/%d} documents successfully processed.",
```
```suggestion
        "%d/%d documents successfully processed.",
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            finding["RVA ID"] = rID
+        else:
+            logging.warning(
+                "Skipping record %d: Unable to extract valid RVA ID from '%s'",
```
Consistency.

```suggestion
                'Skipping record %d: Unable to extract valid RVA ID from "%s"',
```

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +49,292 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            and "findings" not in finding.keys()
+        ):
+            logging.warning(
+                "Skipping record %d. Missing 'RVA ID' or 'NCATS ID' field.", index
```
Consistency.

```suggestion
                'Skipping record %d. Missing "RVA ID" or "NCATS ID" field.', index
```

mcdonnnj commented on README.md:
```diff
@@ -98,7 +98,7 @@ pipenv lock
 ## Contributing ##
 
-We welcome contributions!  Please see [`CONTRIBUTING.md`](CONTRIBUTING.md) for
+We welcome contributions! Please see [`CONTRIBUTING.md`](CONTRIBUTING.md) for
```
We try to avoid changing things from the skeleton unnecessarily.

```suggestion
We welcome contributions!  Please see [`CONTRIBUTING.md`](CONTRIBUTING.md) for
```

mcdonnnj reviewed (changes requested):
Looking at this I do not think it will correctly process v2 schema JSONs. Looking at the sample data I _believe_ the findings data we want is in `json_data["findings"]["individual_findings"]` but we never bring that out during the extraction process. The difference in the schemas require different processing which isn't being done here.

S4lt5 committed:
Apply suggestions from code review

Consistency and error handling suggestions

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

S4lt5 committed:
Update 'schema' to 'Schema' in dictionary key

match previous change.

S4lt5 commented:
> Looking at this I do not think it will correctly process v2 schema JSONs. Looking at the sample data I _believe_ the findings data we want is in `json_data["findings"]["individual_findings"]` but we never bring that out during the extraction process. The difference in the schemas require different processing which isn't being done here.

To be clear, are you saying that the data should be restricted to just the findings and exclude the rest of the JSON object?

I was given pretty minimal guidance here, and my personal thought was to preserve as much of the original structure of what the RVA folks were inputting. The intent was to pass in a single collection with all of the relevant data, knowing that no matter what we do here the format is changing enough that those consuming this data (if any such individuals exist...) will have to adapt to  the new changes.

Much of the other data could be construed as a finding if e.g. there was a successful phishing campaign, weak physical security, etc.... that contributed to the success of a campaign, and I am not sure leaving things out is the best idea, unless this is getting consumed elsewhere. I'd recommend going back to PTP authors and verifying.

S4lt5 committed:
fix signature type hints.

Not sure on how to do advance classes or if it is even worth it (in python), so for e.g. S3 Client, types.Any was used

Also, for extract_findings, the function will accept either a List<dict> or dict object and wrap it accordingly. For consistency, i've added the expected list call, but this may also be called with a dict param.

S4lt5 committed:
Setup type hints for database connection

Handle errors consistently in db connection setup routine (re raise caught execption after logging)

mcdonnnj reviewed (changes requested):
Based on additional conversation around this transition please do the following:

- Remove "v1" schema processing functionality.
- Remove field mapping functionality and documentation as it is no longer necessary.
- Remove any other pre-processing of the JSON data. We should just load the v2 JSON data as-is into the database with no manipulation.
- While we're here please change the file download functionality to just directly load the S3 object into memory instead of using a local temporary file.

mcdonnnj reviewed (changes requested):
Since concerns have been raised about legacy JSONs laying around and retaining the ability to process them my new change request is:

1. Check if the JSON data is a list (v1) or dict (v2) and process accordingly. Something like

```python
if isinstance(findings_data, list):
  process_v1_data()
elif isinstance(findings_data, dict):
  process_v2_data()
else:
  # Log a message, raise an error, etc.
```
2. Process v1 data the same as it currently gets processed (in the `develop` branch) with field mapping etc.
3. Process the v2 data as-is. Do not perform any field mapping or pre-processing.
4. Change S3 object download to pull directly into memory instead of writing to a temporary file.

The v1 schema is not enforced at the database level and has been generally free-form except for massaging the assessment ID to an `RVA ID` key and adjusting the names of three other keys. Since the new schema completely deviates from anything resembling the old schema I see no point in trying to modify any of the data for v2 files being ingested. In fact the two schemas are so dissimilar that having an `RVA ID` key in v1 documents and an `id` key in v2 documents will help differentiate them in the database and I find this preferable to shoehorning a `Schema`/`schema` key into documents. If such a field is desired it should be added in the JSON generating project and has no place in v1 data.

S4lt5 commented:
> Since concerns have been raised about legacy JSONs laying around and retaining the ability to process them my new change request is:
> 
> 1. Check if the JSON data is a list (v1) or dict (v2) and process accordingly. Something like
> 
> ```python
> if isinstance(findings_data, list):
>   process_v1_data()
> elif isinstance(findings_data, dict):
>   process_v2_data()
> else:
>   # Log a message, raise an error, etc.
> ```
> 
> 2. Process v1 data the same as it currently gets processed (in the `develop` branch) with field mapping etc.
> 3. Process the v2 data as-is. Do not perform any field mapping or pre-processing.
> 4. Change S3 object download to pull directly into memory instead of writing to a temporary file.
> 
> The v1 schema is not enforced at the database level and has been generally free-form except for massaging the assessment ID to an `RVA ID` key and adjusting the names of three other keys. Since the new schema completely deviates from anything resembling the old schema I see no point in trying to modify any of the data for v2 files being ingested. In fact the two schemas are so dissimilar that having an `RVA ID` key in v1 documents and an `id` key in v2 documents will help differentiate them in the database and I find this preferable to shoehorning a `Schema`/`schema` key into documents. If such a field is desired it should be added in the JSON generating project and has no place in v1 data. The routine already logs an error if neither is satisfied and the item is skipped, just like the original logic.

I'll concede that NOT doing field mapping on the v2 schema is probably a smart move, but I think that the amount of type checking is already quite un-pythonic (I myself checked to see if an Item was a list or not, which is probably already a bad sign). In general, if its "v1" it's a list, and that can be the breakpoint, so I suppose putting the fieldmapping in the "not a list" branch makes sense.

For 4, if it's a lambda there is temp storage that will just go away anyways, right? With a lack of tests I'd prefer to not touch anything I didn't have to. You can probably just call get_object, but I'd argue that's it's own change and not really relevant to what we are doing here, and since the temp files should be cleaned up between runs, I don't really wanna break something and set us back yet another week.

I think checking if something is a dict is weird for python, but I can at least split the logic for #2 and #3 based on if I get a collection or not which is already there, I'm not really gonna touch 4 as it's not relevant to this change, and adding more nearly 2 months after the original PR was opened seems a little silly to me.

The main intention for breaking out the other functions into their own routines was to allow for some testability, and the intent was just a pure refactoring.

mcdonnnj commented:
> > Since concerns have been raised about legacy JSONs laying around and retaining the ability to process them my new change request is:
> > 
> > 1. Check if the JSON data is a list (v1) or dict (v2) and process accordingly. Something like
> > 
> > ```python
> > if isinstance(findings_data, list):
> >   process_v1_data()
> > elif isinstance(findings_data, dict):
> >   process_v2_data()
> > else:
> >   # Log a message, raise an error, etc.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. Process v1 data the same as it currently gets processed (in the `develop` branch) with field mapping etc.
> > 3. Process the v2 data as-is. Do not perform any field mapping or pre-processing.
> > 4. Change S3 object download to pull directly into memory instead of writing to a temporary file.
> > 
> > The v1 schema is not enforced at the database level and has been generally free-form except for massaging the assessment ID to an `RVA ID` key and adjusting the names of three other keys. Since the new schema completely deviates from anything resembling the old schema I see no point in trying to modify any of the data for v2 files being ingested. In fact the two schemas are so dissimilar that having an `RVA ID` key in v1 documents and an `id` key in v2 documents will help differentiate them in the database and I find this preferable to shoehorning a `Schema`/`schema` key into documents. If such a field is desired it should be added in the JSON generating project and has no place in v1 data. The routine already logs an error if neither is satisfied and the item is skipped, just like the original logic.
> 
> I'll concede that NOT doing field mapping on the v2 schema is probably a smart move, but I think that the amount of type checking is already quite un-pythonic (I myself checked to see if an Item was a list or not, which is probably already a bad sign). In general, if its "v1" it's a list, and that can be the breakpoint, so I suppose putting the fieldmapping in the "not a list" branch makes sense.
> 
> For 4, if it's a lambda there is temp storage that will just go away anyways, right? With a lack of tests I'd prefer to not touch anything I didn't have to. You can probably just call get_object, but I'd argue that's it's own change and not really relevant to what we are doing here, and since the temp files should be cleaned up between runs, I don't really wanna break something and set us back yet another week.
> 
> I think checking if something is a dict is weird for python, but I can at least split the logic for #2 and #3 based on if I get a collection or not which is already there, I'm not really gonna touch 4 as it's not relevant to this change, and adding more nearly 2 months after the original PR was opened seems a little silly to me.
> 
> The main intention for breaking out the other functions into their own routines was to allow for some testability, and the intent was just a pure refactoring.

I'm not sure why you think checking object type is un-Pythonic as it is used regularly in well-known and mature Python projects. It is a necessity when dealing with loosely defined (or generic) inputs such as in this situation. If we are already checking for a `list` for v1 it makes sense to check that we are at least getting a `dict` for v2 as anything else could be a sign of broken process in anything leading up to when the file is retrieved from the S3 bucket. Field mapping should go in the ***is*** a `list` breakpoint as if it is not a `list` we at least know it is unlikely to be v1 data (the only kind of data that should be pre-processed). If you don't want to update to load the S3 object into memory that's fine. I was simply bringing it up so that we could improve things now before this project goes back into the freezer of unmaintained projects as there is still no long-term maintainer for this project.

jmorrowomni commented:
@Yablargo Can you please let me know an estimated timeline of having # 1-3 from https://github.com/cisagov/findings-data-import-lambda/pull/55#pullrequestreview-1244679711 implemented?  Once those are complete, we should be good to go for testing and a push to prod shortly thereafter.

I agree that the last point is probably out of scope for this PR.

S4lt5 commented:
https://github.com/cisagov/findings-data-import-lambda/blob/2c3ed77281388897a149e60c278a42652a40c367/src/findings_data_import.py#L210-L217

Noticed this when looking through for refactor, I think this is causing key errors in the current production version because I don't think it's checking for severity field before trying to key on it.

S4lt5 committed:
Address required changes @ https://github.com/cisagov/findings-data-import-lambda/pull/55

Break out extract_findings, and tag v1/v2 at this stage.
Check for all required fields, and then when committing records, validate according to schema tag.

Tests @ https://github.com/Yablargo/findings-data-import-lambda/tree/new-schema-update-with-tests

S4lt5 commented:
> @Yablargo Can you please let me know an estimated timeline of having # 1-3 from [#55 (review)](https://github.com/cisagov/findings-data-import-lambda/pull/55#pullrequestreview-1244679711) implemented? Once those are complete, we should be good to go for testing and a push to prod shortly thereafter.
> 
> I agree that the last point is probably out of scope for this PR.


@jmorrowomni :
Changes are @ https://github.com/cisagov/findings-data-import-lambda/pull/55/commits/af9a16be198c515e8fbde26cf0e2e191faa3c355

Updated unit tests are at branch https://github.com/Yablargo/findings-data-import-lambda/tree/new-schema-update-with-tests
 I strongly advise against pushing these templates out with no testing, there's clear issues in the original prod code that would have been covered with basic testing.
 
I don't claim that the this PR is absolutely free of bugs, but there would have been a lot more (accidental indentation etc) without em!

jmorrowomni commented on src/findings_data_import.py:
```diff
@@ -49,12 +53,372 @@ def move_processed_file(s3_client, bucket, folder, filename):
+
+
+def validate_v2_findings(findings_data: dict):
+    """Validate a list of V1 findings, discarding invalid entries (such as those with no severity which are not explicitly 'findings').
```
Please update the comment to reflect validating the v2 findings vs the v1 findings

jmorrowomni replied:
Comment was updated in [26198f1](https://github.com/cisagov/findings-data-import-lambda/pull/55/commits/26198f1b8328c746c1d0fa8245eb204adb3280ec), this conversation can be closed @jsf9k

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -20,6 +21,9 @@
 
 SUCCEEDED_FOLDER = "success"
 FAILED_FOLDER = "failed"
+# Identifier for V1 vs V2 schema
+V1_SCHEMA = "v1"
+V2_SCHEMA = "v2"
```
This is due to the other suggestions regarding schema fields.

```suggestion
```

S4lt5 replied:
We're not gonna change this at this time. If it is an issue I suggest we do it as part of a larger overhaul that covers how V1 and V2 *should* work, or after usage for a while we just rip out any "V1" style operations at all since this schema should be phasing out.

jsf9k replied:
Only the person who originally created the conversation (or someone that person directed) should be resolving conversations.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +53,372 @@ def move_processed_file(s3_client, bucket, folder, filename):
+                finding["RVA ID"],
+            )
+            continue
+        # flag this as V1 so update knows how to handle it, and its clear to folks viewing the data downstream
+        finding["Schema"] = V1_SCHEMA
```
This field is not part of the v1 schema so we should not add it to a document.

```suggestion
```

S4lt5 replied:
We're not gonna change this at this time. If it is an issue I suggest we do it as part of a larger overhaul that covers how V1 and V2 *should* work, or after usage for a while we just rip out any "V1" style operations at all since this schema should be phasing out.

jsf9k replied:
Only the person who originally created the conversation (or someone that person directed) should be resolving conversations.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +53,372 @@ def move_processed_file(s3_client, bucket, folder, filename):
+            logging.warning('Skipping record. Missing "id" or "findings" field.')
+            raise ValueError("Missing id or findings field in v2 findings object")
+
+        # Get RVA ID in format DDDD([.-]D+) from the end of the "RVA ID" field.
+        rvaId = re.search(r"(\d{4})(?:[.-](\d+))?$", findings_data["id"])
+        if rvaId:
+            rID = f"RV{rvaId.group(1)}"
+            if rvaId.group(2) is not None:
+                rID += f".{rvaId.group(2)}"
+            findings_data["id"] = rID
+        else:
+            logging.warning(
+                "Skipping record. Unable to extract valid RVA ID from %s",
+                findings_data["id"],
+            )
+            raise ValueError("Unable to parse RVA id from finding object")
```
We should not be modifying anything coming out of a new v2 JSON.

```suggestion
```

S4lt5 replied:
Why is the original field modified? It would be great if you could share the original requirements doc - this is present in the existing logic and I didn't remove it because I don't see any reason to change that specific behavior that is applicable to both schemas.

Looking back in the history, you were involved in the original code that does this, can you explain why the change happens at all, and why it would be different now?

mcdonnnj replied:
Unfortunately there was no original requirements doc to my knowledge and I came in to help the contractor assigned the task toward the end. The `RVA ID` in the v1 schema is changed because (at the time, I don't know about now) the provided assessment ID could include a "phase" e.g. `RV0801.01`. This flattened any such IDs to a single, consistent ID e.g. `RV0801`. This works for the v1 data because all findings are individual documents. It is safe to overwrite a given single finding as it would apply to the same assessment. With v2 even if they were still using the "phase" in the `id` it would **not** be safe to overwrite a document because it could result in the loss of data.

S4lt5 replied:
Makes sense, but should probably check with RVA so they know one way or the other, they may not have considered this at all.

jmorrowomni replied:
@Yablargo Due to the increased chance of overwriting data with the entire file being saved as one entry, please proceed with the change to _not_ modify the `RVA ID` field when processing v2 data.  I'll be sure to coordinate with RVA and inform them of the change and the reasoning behind it.

S4lt5 replied:
@jmorrowomni  I've removed the RVA ID format parsing/validation and it will now take any RVA ID passed in.

jsf9k replied:
Only the person who originally created the conversation (or someone that person directed) should be resolving conversations.

mcdonnnj commented on src/findings_data_import.py:
```diff
@@ -49,12 +53,372 @@ def move_processed_file(s3_client, bucket, folder, filename):
+                findings_data["id"],
+            )
+            raise ValueError("Unable to parse RVA id from finding object")
+        findings_data["Schema"] = V2_SCHEMA
```
If this field is desired it should be defined in the JSON generating project.

```suggestion
```

S4lt5 replied:
We're not gonna change this at this time. If it is an issue I suggest we do it as part of a larger overhaul that covers how V1 and V2 *should* work, or after usage for a while we just rip out any "V1" style operations at all since this schema should be phasing out.

jsf9k replied:
Only the person who originally created the conversation (or someone that person directed) should be resolving conversations.

S4lt5 committed:
Fix comment on _v2_findings processor

S4lt5 committed:
Remove RVA ID validation/parsing as requested

mcdonnnj commented:
I updated my testing environment Lambda with the latest Action run's artifact (Python 3.9 version). I copied the v1 and v2 schema test files from https://github.com/cisagov/pen-test-portal/pull/3 into the appropriate bucket and verified that the two Lambda runs did not fail in the CloudWatch logs.

felddy commented:
This PR does not meet our coding standards and has not passed peer review.  The outstanding deficiencies in the PR are not captured in issues.   There are unresolved conversations and change requests.  This PR is being merged (bypassing branch protections) at the request of the assessment team's leadership against our better judgment.

felddy commented:
This PR does not meet our coding standards and has not passed peer review. The outstanding deficiencies in the PR are not captured in issues. There are unresolved conversations and change requests. This PR is being merged (bypassing branch protections) at the request of the assessment team's leadership against our better judgment.

S4lt5 commented:
> This PR does not meet our coding standards and has not passed peer review. The outstanding deficiencies in the PR are not captured in issues. There are unresolved conversations and change requests. This PR is being merged (bypassing branch protections) at the request of the assessment team's leadership against our better judgment.

Huh? The unresolved issues revolve around changing inbound object schema, which the project has been doing since its origination, ala  https://github.com/cisagov/findings-data-import-lambda/blob/2c3ed77281388897a149e60c278a42652a40c367/src/findings_data_import.py#L186-L207
